[BUG] Retry in case of 412 errors

**Describe the bug**

With Vault 1.10, a 412 status code is returned from Vault if you issue a request to a performance standby node right after obtaining a token via login. The action fails when it sees the 412 status code instead of retrying the request.

See [new-error-response-for-login-requests-to-perf-standbys-lagging-behind-active-node](https://www.vaultproject.io/docs/upgrading/upgrade-to-1.10.x#new-error-response-for-login-requests-to-perf-standbys-lagging-behind-active-node)

**To Reproduce**

Use the vault-action step with jwt authentication against Vault. You must have Vault running with performance standby nodes.

**Expected behavior**

The action should retry the requests that return a 412 status code. See https://www.vaultproject.io/docs/faq/ssct#q-is-there-anything-else-i-need-to-consider-to-achieve-consistency-besides-upgrading-to-vault-1-10

**Log Output**

```
2022-06-13T08:11:34.8806297Z ::group::Get Vault Secrets
2022-06-13T08:11:34.8806631Z ##[group]Get Vault Secrets
2022-06-13T08:11:35.2916908Z ::add-mask::***
2022-06-13T08:11:35.2918818Z ##[debug]Retrieving Vault Token from v1/auth/jwt/login endpoint
2022-06-13T08:11:35.3811815Z ##[debug]✔ Vault Token successfully retrieved
2022-06-13T08:11:35.3812387Z ::group::Token Info
2022-06-13T08:11:35.3812642Z ##[group]Token Info
2022-06-13T08:11:35.3813410Z ##[debug]Operating under policies: ["default","gha-runner"]
2022-06-13T08:11:35.3814198Z ##[debug]Token Metadata: {"role":"github-sm-image-builder"}
2022-06-13T08:11:35.3814672Z ::endgroup::
2022-06-13T08:11:35.3814921Z ##[endgroup]
2022-06-13T08:11:35.3819620Z ::add-mask::***
2022-06-13T08:11:35.4495051Z ::endgroup::
2022-06-13T08:11:35.4495445Z ##[endgroup]
2022-06-13T08:11:35.4523447Z ##[error]Response code 412 (Precondition Failed)
2022-06-13T08:11:35.4554186Z ##[debug]Node Action run completed with exit code 1
2022-06-13T08:11:35.4568168Z ##[debug]VAULT_TOKEN='***'
2022-06-13T08:11:35.4575999Z ##[debug]Finishing: Get Secrets from vault
```

**Additional context**

The `got` library doesn't consider 412 status code as retryable, see https://github.com/sindresorhus/got/blob/HEAD/documentation/7-retry.md#retry

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 ## Unreleased
 
+Bugs:
+
+* Errors due to replication delay for tokens will now be retried [GH-333](https://github.com/hashicorp/vault-action/pull/333)
+
 ## 2.4.1 (April 28th, 2022)
 
 Improvements:

--- a/package.json
+++ b/package.json
@@ -58,6 +58,7 @@
         "@zeit/ncc": "^0.22.3",
         "jest": "^28.1.1",
         "jest-when": "^3.5.1",
+        "mock-http-server": "^1.4.5",
         "semantic-release": "^19.0.3"
     }
 }

--- a/src/action.js
+++ b/src/action.js
@@ -26,7 +26,15 @@ async function exportSecrets() {
     const defaultOptions = {
         prefixUrl: vaultUrl,
         headers: {},
-        https: {}
+        https: {},
+        retry: {
+            statusCodes: [
+                ...got.defaults.options.retry.statusCodes,
+                // Vault returns 412 when the token in use hasn't yet been replicated
+                // to the performance replica queried. See issue #332.
+                412,
+            ]
+        }
     }
 
     const tlsSkipVerify = (core.getInput('tlsSkipVerify', { required: false }) || 'false').toLowerCase() != 'false';

--- a/src/action.test.js
+++ b/src/action.test.js
@@ -8,7 +8,6 @@ const got = require('got');
 const {
     exportSecrets,
     parseSecretsInput,
-    parseResponse,
     parseHeadersInput
 } = require('./action');
 

--- a/src/retries.test.js
+++ b/src/retries.test.js
@@ -1,0 +1,69 @@
+jest.mock('@actions/core');
+
+const core = require('@actions/core');
+const ServerMock = require("mock-http-server");
+const { exportSecrets } = require("./action");
+const { when } = require('jest-when');
+
+describe('exportSecrets retries', () => {
+    var server = new ServerMock({ host: "127.0.0.1", port: 0 });
+    var calls = 0;
+
+    beforeEach((done) => {
+        calls = 0;
+        jest.resetAllMocks();
+
+        when(core.getInput)
+            .calledWith('token', expect.anything())
+            .mockReturnValueOnce('EXAMPLE');
+
+        when(core.getInput)
+            .calledWith('secrets', expect.anything())
+            .mockReturnValueOnce("kv/mysecret key");
+
+        server.start(() => {
+            expect(server.getHttpPort()).not.toBeNull();
+            when(core.getInput)
+                .calledWith('url', expect.anything())
+                .mockReturnValueOnce('http://127.0.0.1:' + server.getHttpPort());
+            done();
+        });
+    });
+
+    afterEach((done) => {
+        server.stop(done);
+    });
+
+    function mockStatusCodes(statusCodes) {
+        server.on({
+            path: '/v1/kv/mysecret',
+            reply: {
+                status: function() {
+                    let status = statusCodes[calls];
+                    calls += 1;
+                    return status;
+                },
+                headers: { "content-type": "application/json" },
+                body: function() {
+                    return JSON.stringify({ data: {"key": "value"} })
+                }
+            }
+        });
+    }
+
+    it('retries on 412 status code', (done) => {
+        mockStatusCodes([412, 200])
+        exportSecrets().then(() => {
+            expect(calls).toEqual(2);
+            done();
+        });
+    });
+
+    it('retries on 500 status code', (done) => {
+        mockStatusCodes([500, 200])
+        exportSecrets().then(() => {
+            expect(calls).toEqual(2);
+            done();
+        });
+    });
+});

